libursa/src/bn/rust.rs

```diff
@@ -0,0 +1,632 @@
+use errors::UrsaCryptoError;
+
+use hash::{digest, DigestAlgorithm, Digest, sha2};
+use num_bigint::{BigInt, BigUint, RandBigInt, ToBigInt, Sign};
+use num_integer::Integer;
+use num_traits::identities::{One, Zero};
+use num_traits::{Num, Signed, ToPrimitive, Pow};
+use glass_pumpkin::{prime, safe_prime};
+use rand::rngs::OsRng;
+
+#[cfg(feature = "serialization")]
+use serde::ser::{Serialize, Serializer, Error as SError};
+
+#[cfg(feature = "serialization")]
+use serde::de::{Deserialize, Deserializer, Visitor, Error as DError};
+
+use std::error::Error;
+use std::fmt;
+use std::cmp::Ord;
+use std::cmp::Ordering;
+
+pub struct BigNumberContext;
+
+pub struct BigNumber {
+    bn: BigInt
+}
+
+impl BigNumber {
+    pub fn new_context() -> Result<BigNumberContext, UrsaCryptoError> {
+        Ok(BigNumberContext{})
+    }
+
+    pub fn new() -> Result<BigNumber, UrsaCryptoError> {
+        Ok(BigNumber {
+            bn: BigInt::zero()
+        })
+    }
+
+    pub fn generate_prime(size: usize) -> Result<BigNumber, UrsaCryptoError> {
+        match prime::new(size)?.to_bigint() {
+            Some(bn) => Ok(BigNumber { bn }),
+            None => Err(UrsaCryptoError::InvalidStructure("Unable to generate prime".to_string()))
+        }
+    }
+
+    pub fn generate_safe_prime(size: usize) -> Result<BigNumber, UrsaCryptoError> {
+        match safe_prime::new(size)?.to_bigint() {
```

Comment:
```
match prime::new(size)?.to_bigint() {
            Some(bn) => Ok(BigNumber { bn }),
            None => Err(UrsaCryptoError::InvalidStructure("Unable to generate prime".to_string()))
        }
```
This code can be encapsulated in a macro that takes a prime generator, i.e `prime` or `safe_prime` thus avoiding code duplication